### ISSUE TITLE
Background processes make stdin invisible

### ISSUE THREAD

**Issue by Granitosaurus** [xonsh/xonsh#1368]
Tested in v0.4.2 and ipython 4.1.2 and python 3.5

```
$ python
>> "foo"
>> Ctrl+z
$ python
>> <invisible text>
>> Ctrl+z or even Ctrl+d
$ fg python(first one)
>> "foo" # everything is visible here
>> Ctrl+z
$ killall python
# every stdin from this point is invisible and only a restart of xonsh can fix it
# seems like the original job with stdin takes the stdin to the grave with it :D
```

Even though the text is invisible it is still there and is interpreted as it would be normally.
Now it seems that the only way to resolve this is to restart xonsh shell completely since killing of the ipython  processes doesn't resolve the issue.

Edit: this is also an issue with standard python shell
Edit: It appears that stdin in `cat` is visible but is all messed up, like it doesn't allow you to delete characters or insert newlines.

**Comment by scopatz** [comment]
Thanks for reporting @Granitas. The same basic thing happens to me in xonsh and Bash with the hiding of stdin.  Conceptually, I am not sure that it is well defined to show stdin on a background process (it is in the background) because its special file descriptors for stdin, stdout, and stderr are all being redirected.

What is the behaviour you were expecting and what changes can be made?  Sorry I haven't thought a lot about this corner case

**Comment by Granitas** [comment]
@scopatz The thing here is that a suspended job shouldn't block stdin. I'm I'm not having this issue on bash or zsh and I'm not sure but I think this bug didn't exist in Xonsh 0.3.

**Comment by scopatz** [comment]
Ahh that is fair. This is almost certainly a job control issue then.  Would you mind trying to find the commit that broke the behaviour for you with git bisect?  That would really help!

**Comment by PaulReiber** [comment]
if the characters from stdin are being consumed and processed just fine,
but are "invisible" that means they're not getting echoed out onto the
screen as you type them.

That's an aspect of the terminal's mode.  When you control-Z there's no
telling what the terminal mode will be left in as the process suspends, and
usually, the process being suspended is pretty stupid about cleaning up
after itself.  Some do better than others - vi and emacs reset the terminal
mode nicely when backgrounding, for example.  But not all programs play
fair, so the xonsh shell needs to ensure it's setting the mode right in
these case, rather than leaving things to chance.

Kind regards,


*Paul Reiber*http://about.me/reiber

**Comment by scopatz** [comment]
Pull requests very welcome here.

**Comment by PaulReiber** [comment]
Anthony can you please check your gmail spam filter? wondering if a few messages I've emailed your way have maybe been blocked.

**Comment by PaulReiber** [comment]
I've identified a probable fix for this.  sys.stdin's termio structure needs ECHO toggled on just in case it somehow got toggled off.

Looks like the fix would go into readline_shell.py (and maybe base_shell.py but not sure)

https://docs.python.org/2/library/termios.html has an example at the end of the page
which toggles  "ECHO" on sys.stdin prior to prompting for a password.  It's a bad example in that it assumes that ECHO's on at first, but hey.

```python
def getpass(prompt="Password: "):
    import termios, sys
    fd = sys.stdin.fileno()
    old = termios.tcgetattr(fd)
    new = termios.tcgetattr(fd)
    new[3] = new[3] & ~termios.ECHO          # lflags
    try:
        termios.tcsetattr(fd, termios.TCSADRAIN, new)
        passwd = raw_input(prompt)
    finally:
        termios.tcsetattr(fd, termios.TCSADRAIN, old)
    return passwd
```

So we would need something like this:

```python
import termios, sys

    fd = sys.stdin.fileno()
    si_attrs = termios.tcgetattr(fd)
    si_attrs[3] = si_attrs[3] & termios.ECHO
    termios.tcsetattr(fd, termios.TCSADRAIN, si_attrs)
```

**Comment by PaulReiber** [comment]
Sharing what I've learned so far.

Goal: turning back on self.stdin's ECHO attribute via a termios.tcsetattr() call at just the right time - either right before or right after printing the prompt, when we're about to read more input characters from the end user.   (i.e. 2nd arg should be termios.TCSANOW probably)

Forcing a traceback helped me to identify _cmdloop() as a good candidate.

```python
(venv)pbr@mongo:~/xonsh/xonsh$ XONSH_DEBUG=1 xonsh
(venv) pbr@mongo ~/xonsh/xonsh master $ import traceback
(venv) pbr@mongo ~/xonsh/xonsh master $ traceback.print_stack()
  File "/home/pbr/xonsh/venv/bin/xonsh", line 3, in <module>
    main()
  File "/home/pbr/xonsh/venv/lib/python3.4/site-packages/xonsh/main.py", line 219, in main
    shell.shell.cmdloop()
  File "/home/pbr/xonsh/venv/lib/python3.4/site-packages/xonsh/readline_shell.py", line 376, in cmdloop
    self._cmdloop(intro=intro)
  File "/home/pbr/xonsh/venv/lib/python3.4/site-packages/xonsh/readline_shell.py", line 362, in _cmdloop
    stop = self.onecmd(line)
  File "/usr/lib/python3.4/cmd.py", line 211, in onecmd
    return self.default(line)
  File "/home/pbr/xonsh/venv/lib/python3.4/site-packages/xonsh/base_shell.py", line 165, in default
    run_compiled_code(code, self.ctx, None, 'single')
  File "/home/pbr/xonsh/venv/lib/python3.4/site-packages/xonsh/codecache.py", line 66, in run_compiled_code
    func(code, glb, loc)
  File "<xonsh-code>", line 1, in <module>
```

I've tried various things including some interactive debugging.  So far, no joy.  Not sure why the call I'm making to termios.tcsetattr isn't turning back on "echo".

Anyone?  Input's welcome.

**Comment by scopatz** [comment]
Thanks for the info @PaulReiber, would you mind putting in a PR for whatever you have.  I think it is easier to work from concrete code.

**Comment by scopatz** [comment]
I believe this has been addressed by #2123. Please feel free to reopen.
